tests/components/samsungtv/test_media_player.py

```diff
@@ -55,247 +55,316 @@ class UnhandledResponse(Exception):
+    """Test setup of platform with no data."""
+    with mock.patch("homeassistant.components.samsungtv.media_player.socket"):
+        with patch(
+            "homeassistant.components.samsungtv.media_player.LOGGER.warning"
+        ) as mocked_warn:
             add_entities = mock.Mock()
-            setup_platform(self.hass, {}, add_entities, discovery_info=None)
+            await async_setup_platform(hass, {}, add_entities, discovery_info=None)
             mocked_warn.assert_called_once_with("Cannot determine device")
             add_entities.assert_not_called()
 
-    def test_update_on(self):
-        """Testing update tv on."""
-        self.device.update()
-        self.assertEqual(STATE_ON, self.device._state)
 
-    def test_update_off(self):
-        """Testing update tv off."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(side_effect=OSError("Boom"))
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        self.device.update()
-        assert STATE_OFF == self.device._state
-
-    def test_send_key(self):
-        """Test for send key."""
-        self.device.send_key("KEY_POWER")
-        self.assertEqual(STATE_ON, self.device._state)
-
-    def test_send_key_broken_pipe(self):
-        """Testing broken pipe Exception."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(side_effect=BrokenPipeError("Boom"))
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        self.device.send_key("HELLO")
-        self.assertIsNone(self.device._remote)
-        self.assertEqual(STATE_ON, self.device._state)
-
-    def test_send_key_connection_closed_retry_succeed(self):
-        """Test retry on connection closed."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(
-            side_effect=[
-                self.device._exceptions_class.ConnectionClosed("Boom"),
-                mock.DEFAULT,
-            ]
-        )
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        command = "HELLO"
-        self.device.send_key(command)
-        self.assertEqual(STATE_ON, self.device._state)
-        # verify that _remote.control() get called twice because of retry logic
-        expected = [mock.call(command), mock.call(command)]
-        assert expected == _remote.control.call_args_list
-
-    def test_send_key_unhandled_response(self):
-        """Testing unhandled response exception."""
-        _remote = mock.Mock()
-        _remote.control = mock.Mock(
-            side_effect=self.device._exceptions_class.UnhandledResponse("Boom")
-        )
-        self.device.get_remote = mock.Mock(return_value=_remote)
-        self.device.send_key("HELLO")
-        self.assertIsNone(self.device._remote)
-        self.assertEqual(STATE_ON, self.device._state)
+async def test_setup_method_selection():
+    """Test of method selection."""
+    conf = WORKING_CONFIG.copy()
+    device = SamsungTVDevice(**conf)
```

Comment:
Don't create the entity directly. We'll check that the entity behaves correctly by checking the state of the entity after it has been added to home assistant by the platform. And also by calling media_player services and checking that the patched library method are called as we expect.

Reply:
I understand that you're interested in clean code and good coverage, but I have just added some minor changes and now have to fix all the debts that were in the code since months. At the end of the day this will be a complete rewrite of the whole test file. Your comments may be simple to implement for an experienced long time supporter of HA but for me this is leads always to hours of reading and investigation what you actual want me to do. Additionally for 10 checked other integration tests I could found 15 different solutions. So how to know which is the currently correct way?
I'll do my best to satisfy you comments but this really doesn't encourage me to do more PRs. Sorry to say this...

Reply:
If we fix the tests to make them more robust it will be a lot easier to add the other changes going forward. It's best to fix the tests in this first PR. Then hopefully we don't need to change them much later, just add new tests for the new functionality.

I understand it's hard work now, but it will pay off later.